### PR TITLE
formalize audio announcement logic

### PR DESCRIPTION
**Pull request by panentheos**
#### Summary of changes
**Asana Ticket:** [RTS Refactor: Audio Logic Revision](https://app.asana.com/0/1185117109217413/1205450681869477/f)

This overhauls the logic for immediate audio announcements, creating an explicit separation between announcements and passive readouts, removing some edge cases, and changing some announcement behavior. Details:
* Announcements now use their own code path with separate logic (though they still share the various `from_message` calls in the audio modules).
* Behavior change: Headway messages no longer trigger immediate announcements.
* Behavior change: When predictions appear after not being present previously (e.g. because of a headway or custom message), the first prediction will be announced immediately.
* Due to this new announcement type, many signs will make a single prediction announcement when the app is restarted.
* Creation of `Content.Audio` structs have been standardized in `from_message` calls.
* Mezzanine signs now behave consistently, and both sides can generate their own independent announcements.
* Light rails announcements only include the relevant prediction, matching the behavior of heavy rail signs.
* Audios are no longer sorted based on type. This theoretically results in a behavior change, but due to how audio readouts are timed and scheduled, it would be extremely rare to see this in practice.

### CODE REVIEW

File: lib/signs/utilities/audio.ex [hunk 1]
```diff
@@ -11,207 +11,220 @@ defmodule Signs.Utilities.Audio do
...
         end
       end)
 
-    {new_audios, sign}
+    {audios ++ new_audios, sign}
+  end
+
+  defp cache_value(list, value), do: [value | list] |> Enum.take(@announced_history_length)
+
+  defp decode_sign(sign) do
```
Comment (panentheos) [inline lib/signs/utilities/audio.ex:153]: Note: This function provides a sneak peek into where this refactor is headed. It's essentially "decompiling" the granular line-oriented messages into higher-level semantic structures, which are more suitable for transforming and generating e.g. audio messages. Eventually, the goal is to make proper types for these, and create them earlier in the process, so the semantic forms can be used more naturally, without having to reverse engineer them.

**Review by PaulJKim — COMMENTED** [review]
Re: the note `Light rails announcements only include the relevant prediction, matching the behavior of heavy rail signs`

Does "relevant prediction" in this case just refer to the soonest prediction? If so, is it possible we'd want to do additional immediate announcements in some cases for light rail when a platform is serving multiple branches? e.g. a rider might be waiting at Kenmore for a D train when the sign changes and triggers an immediate readout, but the first prediction is for a C train and the second prediction for a D train, so the rider would need to wait until a passive readout to hear about the D train even though for that rider, the D train prediction is arguably the relevant one

File: lib/signs/utilities/audio.ex [hunk 2]
```diff
@@ -238,18 +251,6 @@ defmodule Signs.Utilities.Audio do
...
-    Enum.sort_by(audios, fn audio ->
-      case audio do
-        %Content.Audio.TrainIsBoarding{} -> 1
-        %Content.Audio.TrainIsArriving{} -> 2
-        %Content.Audio.Approaching{} -> 3
-        _ -> 4
-      end
-    end)
-  end
-
```
[the remark this answers is not in the document]
Comment (panentheos) [reply lib/signs/utilities/audio.ex:252]: Even in the old code, an arriving message would suppress the readout of a normal prediction, so that particular situation wouldn't happen. However, an important point is that this sorting only happens within the context of a simultaneous update. If the statuses came in 1 second apart, the first one to change would get read first, regardless. That's why I think this isn't something we should worry about.

File: lib/signs/utilities/audio.ex [hunk 3]
```diff
@@ -11,207 +11,220 @@ defmodule Signs.Utilities.Audio do
...
-          )
-
-          true
+  defp get_custom_announcements({audios, sign}, items) do
+    case Enum.find(items, &match?({:custom, _, _}, &1)) do
+      {:custom, top, bottom} ->
+        [audio] = Audio.Custom.from_messages(top, bottom)
+
+        if sign.announced_custom_text != audio.message do
+          {audios ++ [audio], %{sign | announced_custom_text: audio.message}}
```
[the remark this answers is not in the document]
Comment (panentheos) [reply lib/signs/utilities/audio.ex:30]: No, the old algorithm did immediate readouts of new custom messages too.

File: lib/signs/utilities/audio.ex [hunk 4]
```diff
@@ -11,207 +11,220 @@ defmodule Signs.Utilities.Audio do
...
   end
 
-  def should_interrupting_read?(%Content.Message.Alert.DestinationNoService{}, _sign, _line) do
-    false
-  end
+  defp get_alert_announcements({audios, sign}, items) do
+    case Enum.find(items, &match?({:alert, _, _}, &1)) do
+      {:alert, top, bottom} ->
+        new_audios =
+          case top do
```
[the remark this answers is not in the document]
Comment (panentheos) [reply lib/signs/utilities/audio.ex:44]: Actually, I was imagining the opposite: Once we're no longer tinkering with the audio structs after the fact, the need for an abstract layer there kind of goes away, and we can just think about it as a function that takes sign context and returns audio messages.

File: lib/signs/utilities/audio.ex [hunk 5]
```diff
@@ -11,207 +11,220 @@ defmodule Signs.Utilities.Audio do
...
+
+            match?(%Message.StoppedTrain{}, message) && index == 0 &&
+                {message.trip_id, message.stops_away} not in sign.announced_stalls ->
+              {Audio.StoppedTrain.from_message(message),
+               update_in(
+                 sign.announced_stalls,
+                 &cache_value(&1, {message.trip_id, message.stops_away})
+               )}
+
+            match?(%Message.Predictions{}, message) && is_integer(message.minutes) && index == 0 &&
```
[the remark this answers is not in the document]
Comment (panentheos) [reply lib/signs/utilities/audio.ex:117]: There are two levels of nesting at play to make this work. For platform signs, the data looks like: `[{:predictions, [%P{}, %P{}]}]`, and for mezzanine signs it looks like `[{:predictions, [%P{}]}, {:predictions, [%P{}]}]`. The `index` is for the inner list, so the bottom message of the mezzanine sign will also have index 0, and this clause should trigger.

I know this function in particular is hairy. Do you think some targeted comments here (or elsewhere, perhaps `decode_sign`) would help?

**Comment by panentheos** [comment]
Regarding the light rail question, I think there's a larger discussion we should have around it. I carried forward the existing behavior of announcements delaying passive readouts (actually for _longer_ than the read timer itself, since it adds), but that can definitely cause normal predictions not to be read in certain cases. It's not a simple light/heavy rail issue, either. The point you mentioned applies equally to the RL branches. Also, most mezzanine signs show multiple destinations, so you could make a similar argument there.

My thought was to do the quick standardization for now, and then start the larger conversation (which may lead to further changes), but I'm curious of your thoughts on that plan.

File: lib/signs/utilities/audio.ex (hunk 5 above)
[the remark this answers is not in the document]
Comment (PaulJKim) [reply lib/signs/utilities/audio.ex:117]: Ah I see, didn't catch the distinction between the shape of the data for the platform vs mezzanine cases.

And yeah I was thinking some documentation could be helpful. Although this function is fairly verbose, I actually don't think it's terribly hard to follow or anything. But I think some documentation or possibly even just a typespec could be helpful for the `decode_sign` logic. I think the hard part while reading/reviewing was keeping in mind the different possible outcomes from `decode_sign` while thinking through the rest of the logic

File: lib/signs/utilities/audio.ex (hunk 1 above)
Comment (PaulJKim) [reply lib/signs/utilities/audio.ex:153]: Clarifying question - when you say "make proper types", do you just mean to more explicitly define them using typespecs somewhere or do you mean something else?

File: lib/signs/utilities/audio.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (PaulJKim) [reply lib/signs/utilities/audio.ex:252]: Ah okay, I see what you mean now

**Comment by PaulJKim** [comment]
Hmm yeah I think the thought to do the standardization first makes sense although I'd guess that we'd want to do the standardization in the other direction though if anything. Personally, it seems like the second prediction is still pertinent to riders in many situations. But like you said, that seems to be suggesting a larger discussion about audio behavior. Just to make sure I'm understanding your initial bullet point though - is the light rail behavior change you called out that immediate announcements of predictions will only include the next train and not the following train?

File: lib/signs/utilities/audio.ex (hunk 1 above)
Comment (panentheos) [reply lib/signs/utilities/audio.ex:153]: I meant creating named structs. The data here is already kind of leaning in that direction (with the e.g. `{:predictions, ...}` shapes), but since the scope of this data is very limited to start, I didn't want to jump ahead and make it concrete before we could see more of the use cases.

**Comment by panentheos** [comment]
Correct, light rail prediction announcements now only announce the prediction that is approaching, arriving, or boarding, and not the (following) numeric one. Though if you happen to get multiple at once, e.g. two BRD predictions on a multi-berth platform, then it will announce both, as you'd expect (I'm not sure that situation would actually happen in practice).

File: lib/signs/utilities/audio.ex (hunk 1 above)
Comment (PaulJKim) [reply lib/signs/utilities/audio.ex:153]: Right - considering structs are just maps with a `__struct__` key, you could say these semantic structures are almost there. So the idea is to codify our content types at the high level with named structs, but eliminate the use of types at the end of the pipeline to streamline the serialization process. Am I understanding that correctly?

**Review by PaulJKim — APPROVED** [review]
Tested locally and looks good to me! I think this is heading things in a good direction 👍
